### PR TITLE
Hs 3015 as

### PR DESCRIPTION
**Pull request by aslobodeniuk**

### CODE REVIEW

**Comment by jmcatena** [comment]
As agreed during the daily, I am creating a branch HS-3027 based in this one to incorporate necessary fixes. Please do not merge this to sdk-10.0.23-hls before HS-3027 is merged into this.
Please let me know any concern.

**Comment by aslobodeniuk** [comment]
just a reminder, that this branch has build problems, when we're building for the first time. Something related to accessibility of gstamcmacro.h . Maybe it's not a problem actually.

File: sys/androidmedia/gstamcvideodec.c
Comment (aslobodeniuk) [inline sys/androidmedia/gstamcvideodec.c]: This is a small, but still a performance fall: we're creating empty "video/x-amc" caps each time (it adds several malloc/free), where we could just create it once.

File: sys/androidmedia/gstamcvideodec.c
Comment (aslobodeniuk) [inline sys/androidmedia/gstamcvideodec.c]: This line is unacceptable: we are triggering new "set caps" proccess with each new buffer.

Comment (aslobodeniuk) [reply sys/androidmedia/gstamcvideodec.c]: yes, this line has to be resolved, it's just caused by copying code in a hurry. I did it actually..

File: gst-libs/gst/androidjni/gstjniamcutils.c
Comment (aslobodeniuk) [inline gst-libs/gst/androidjni/gstjniamcutils.c]: "const GstCaps * caps" would be nice

File: gst-libs/gst/androidjni/gstjniamcutils.c
Comment (aslobodeniuk) [inline gst-libs/gst/androidjni/gstjniamcutils.c]: "const GstCaps * caps" would be nice

File: gst-libs/gst/androidjni/gstjniamcutils.c
Comment (aslobodeniuk) [inline gst-libs/gst/androidjni/gstjniamcutils.c]: do we really need this trace? this seems to be a "development one"

Comment (palegre-fluendo) [reply gst-libs/gst/androidjni/gstjniamcutils.c]: Of course it is not critical, we can remove it or set DEBUG level.

File: gst-libs/gst/androidjni/gstjniamcutils.c
Comment (jmcatena) [inline gst-libs/gst/androidjni/gstjniamcutils.c:54]: style: mixed braces and no braces with single statement blocks.

File: gst-libs/gst/androidjni/gstjnimediacodeclist.c
Comment (jmcatena) [inline gst-libs/gst/androidjni/gstjnimediacodeclist.c]: Nit: we have already a function to avoid repeating this block everywhere: j_find_class
Also, gstamc keeps a cache of used classes and methods so that they do not have to be obtained everytime. That probably should belong to a lower level common jni support library, but anyway it's there to be used.

File: gst-libs/gst/androidjni/gstjniutils.h
Comment (jmcatena) [inline gst-libs/gst/androidjni/gstjniutils.h:88]: NIT: Jumping to labels from macros makes usage of this macros very tricky. We should consider another way to communicate the success/failure, like a bool variable as first parameter.

File: sys/androidmedia/gstamc.c
Comment (jmcatena) [inline sys/androidmedia/gstamc.c:58]: NIT: We may want to have the cached jni classes and methods in a more general library, maybe gstgniutils.
Instead of definining each class in a root structure, we could beter use nested structures resembling the java namespaces.

File: sys/androidmedia/gstamc.c
Comment (jmcatena) [inline sys/androidmedia/gstamc.c]: Delete this line, I left it from testing and is not needed.

File: sys/androidmedia/gstamc.c
Comment (jmcatena) [inline sys/androidmedia/gstamc.c]: I think the intention was to check codec_info_class, not codec_list.

File: sys/androidmedia/gstamc.c
Comment (jmcatena) [inline sys/androidmedia/gstamc.c:2946]: Important: In sdk we are raising or lowering codecs supporting tunneled rank by 1 depending on if tunneled is desired or not.
Lowering the rank seems unnecessary and sdk may need to alter rank by larger amount to compensate this.

File: sys/androidmedia/gstamcvideodec.h
Comment (jmcatena) [inline sys/androidmedia/gstamcvideodec.h]: Unused, can be safely removed, the one in _GstAmcCodec is used instead.

**Comment by jmcatena** [comment]
I finished reviewing and I have an important concern:
Currently we are activating adaptive if the decoder supports it, regardless of tunneled or not.
But adaptive requires it to be configured to output to a surface, it does not work when output is byte buffers.
If we are using byte output in any use case, there are many more fixes to do, including those that cause black gaps and issues when changing format.
I suppose that it is not using byte buffer output even in non tunneled mode because adaptive works fine and it should not otherwise, but reviewing the code it's difficult to be sure, so maybe someone that know better can clarify and save more study of this already huge PR.
We cannot keep including big fixes in this PR, we need to merge what's done, but we need to understand what to expect from this state and what's pending if any.

**Review by jmcatena — APPROVED** [review]
I have concerns expressed before, but these may be adressed if necessary in further PRs and are not a reason to block this one.

File: sys/androidmedia/gstamc.c
Comment (ylatuya) [reply sys/androidmedia/gstamc.c:2946]: This is required because otherwise you would have 3 decoders with same rank and by default we want the regular decoder to be used

File: gst-libs/gst/androidjni/gstjniutils.h
Comment (aslobodeniuk) [reply gst-libs/gst/androidjni/gstjniutils.h:88]: I have the way-around opinion: this pattern of having an "error" label in each function should be used in all the code we write, because it simplifies the code, and at the same way forces the code to check and log every unexpected case.

File: gst-libs/gst/androidjni/gstjnimediacodeclist.c
Comment (aslobodeniuk) [reply gst-libs/gst/androidjni/gstjnimediacodeclist.c]: we actually also have gst_jni_get_class, which is the same as j_find_class

**Pull request merged**
